File: frontend/src/presenters/flipOpportunityTablePresenter.test.ts
import { describe, expect, it } from 'vitest';
import type { CurrencyAmount, FlipOpportunity, Technique } from '../entities/FlipOpportunity';
import {
  buildDisplayGroups,
  filterByTechnique,
  filterByThresholds,
  getRouteKey,
  partitionFavorites,
  sortOpportunities,
} from './flipOpportunityTablePresenter';

function currencyAmount(currencyId: string, name: string, quantity: number): CurrencyAmount {
  return { currencyId, name, iconUrl: null, itemType: 'CURRENCY', quantity };
}

function chaosProfit(quantity: number): CurrencyAmount {
  return currencyAmount('chaos', 'Chaos Orb', quantity);
}

function opportunity(overrides: Partial<FlipOpportunity> = {}): FlipOpportunity {
  return {
    technique: 'EXCHANGE_SPREAD',
    start: [currencyAmount('chaos', 'Chaos Orb', 1)],
    via: [currencyAmount('wisdom', 'Scroll of Wisdom', 366)],
    sell: [currencyAmount('chaos', 'Chaos Orb', 1.9784)],
    marginPercent: 97.84,
    profit: chaosProfit(0.9784),
    volume: 1234,
    detail: 'instant 185:1 · competitive 366:1',
    ...overrides,
  };
}

const allEnabled: Record<Technique, boolean> = {
  VENDOR_RECIPE: true,
  EXCHANGE_SPREAD: true,
  DIVINATION_CARD: true,
  BULK_BUY: true,
};

describe('getRouteKey', () => {
  it('is identical for two opportunities with the same technique and currencies', () => {
    const a = opportunity();
    const b = opportunity({ marginPercent: 12, profit: chaosProfit(0.1), volume: 5 });

    expect(getRouteKey(a)).toBe(getRouteKey(b));
  });

  it('differs when the technique differs', () => {
    const a = opportunity({ technique: 'EXCHANGE_SPREAD' });
    const b = opportunity({ technique: 'BULK_BUY' });

    expect(getRouteKey(a)).not.toBe(getRouteKey(b));
  });

  it('differs when the via currency differs', () => {
    const a = opportunity();
    const b = opportunity({
      via: [currencyAmount('portal', 'Portal Scroll', 10)],
    });

    expect(getRouteKey(a)).not.toBe(getRouteKey(b));
  });
});

describe('filterByTechnique', () => {
  it('keeps only opportunities whose technique is enabled', () => {
    const spread = opportunity({ technique: 'EXCHANGE_SPREAD' });
    const bulk = opportunity({ technique: 'BULK_BUY' });

    const result = filterByTechnique([spread, bulk], { ...allEnabled, BULK_BUY: false });

    expect(result).toEqual([spread]);
  });
});

describe('filterByThresholds', () => {
  it('keeps opportunities at or above every provided threshold', () => {
    const low = opportunity({ marginPercent: 10, volume: 5 });
    const high = opportunity({ marginPercent: 90, volume: 500 });

    expect(filterByThresholds([low, high], { margin: 50 })).toEqual([high]);
    expect(filterByThresholds([low, high], { volume: 5 })).toEqual([low, high]);
  });

  it('applies multiple thresholds together', () => {
    const meetsMarginOnly = opportunity({ marginPercent: 90, volume: 1 });
    const meetsBoth = opportunity({ marginPercent: 90, volume: 100 });

    const result = filterByThresholds([meetsMarginOnly, meetsBoth], { margin: 50, volume: 50 });

    expect(result).toEqual([meetsBoth]);
  });

  it('returns everything when no thresholds are provided', () => {
    const items = [opportunity(), opportunity({ marginPercent: 1 })];

    expect(filterByThresholds(items, {})).toEqual(items);
  });
});

describe('sortOpportunities', () => {
  it('sorts ascending by the requested column', () => {
    const low = opportunity({ profit: chaosProfit(1) });
    const high = opportunity({ profit: chaosProfit(9) });

    expect(sortOpportunities([high, low], 'profit', 'asc')).toEqual([low, high]);
  });

  it('sorts descending by the requested column', () => {
    const low = opportunity({ volume: 1 });
    const high = opportunity({ volume: 9 });

    expect(sortOpportunities([low, high], 'volume', 'desc')).toEqual([high, low]);
  });

  it('does not mutate the input array', () => {
    const items = [opportunity({ marginPercent: 1 }), opportunity({ marginPercent: 9 })];
    const original = [...items];

    sortOpportunities(items, 'margin', 'desc');

    expect(items).toEqual(original);
  });
});

describe('partitionFavorites', () => {
  it('splits opportunities into favorites and others by route key', () => {
    const favorited = opportunity({ technique: 'EXCHANGE_SPREAD' });
    const notFavorited = opportunity({ technique: 'BULK_BUY' });

    const result = partitionFavorites([favorited, notFavorited], new Set([getRouteKey(favorited)]));

    expect(result).toEqual({ favorites: [favorited], others: [notFavorited] });
  });

  it('preserves relative order within each group', () => {
    const favA = opportunity({ technique: 'EXCHANGE_SPREAD', volume: 1 });
    const favB = opportunity({
      technique: 'EXCHANGE_SPREAD',
      volume: 2,
      via: [currencyAmount('x', 'X', 1)],
    });
    const keys = new Set([getRouteKey(favA), getRouteKey(favB)]);

    const result = partitionFavorites([favA, favB], keys);

    expect(result.favorites).toEqual([favA, favB]);
  });
});

describe('buildDisplayGroups', () => {
  it('filters by technique and threshold, sorts, then partitions favorites -- a filtered-out favorite disappears entirely', () => {
    const keptFavorite = opportunity({ technique: 'EXCHANGE_SPREAD', marginPercent: 90, volume: 1 });
    const filteredOutFavorite = opportunity({
      technique: 'BULK_BUY',
      marginPercent: 5,
      volume: 1,
      via: [currencyAmount('y', 'Y', 1)],
    });
    const nonFavorite = opportunity({
      technique: 'EXCHANGE_SPREAD',
      marginPercent: 50,
      volume: 2,
      via: [currencyAmount('z', 'Z', 1)],
    });

    const result = buildDisplayGroups([keptFavorite, filteredOutFavorite, nonFavorite], {
      enabledTechniques: { ...allEnabled, BULK_BUY: false },
      thresholds: {},
      sortColumn: 'margin',
      sortDirection: 'desc',
      favoriteRouteKeys: new Set([getRouteKey(keptFavorite), getRouteKey(filteredOutFavorite)]),
    });

    expect(result.favorites).toEqual([keptFavorite]);
    expect(result.others).toEqual([nonFavorite]);
  });
});

File: app/test/poe_flip_finder/flip_opportunity_table_presenter_test.exs
defmodule PoeFlipFinder.FlipOpportunityTablePresenterTest do
  use ExUnit.Case, async: true

  alias PoeFlipFinder.{Currency, CurrencyAmount, FlipOpportunity, FlipOpportunityTablePresenter}

  # Ported 1:1 from flipOpportunityTablePresenter.test.ts.

  defp currency_amount(external_id, display_name, quantity, category \\ :currency) do
    %CurrencyAmount{
      currency: %Currency{
        id: nil,
        external_id: external_id,
        display_name: display_name,
        category: category
      },
      quantity: quantity
    }
  end

  defp chaos_profit(quantity), do: currency_amount("chaos", "Chaos Orb", quantity)

  defp opportunity(overrides \\ %{}) do
    Map.merge(
      %FlipOpportunity{
        technique: :exchange_spread,
        start: [currency_amount("chaos", "Chaos Orb", 1)],
        via: [currency_amount("wisdom", "Scroll of Wisdom", 366)],
        sell: [currency_amount("chaos", "Chaos Orb", 1.9784)],
        margin_percent: 97.84,
        profit: chaos_profit(0.9784),
        start_chaos_equivalent: 1.0,
        volume: 1234,
        detail: "buy 365:1 · sell 186:1"
      },
      overrides
    )
  end

  @all_enabled %{
    vendor_recipe: true,
    exchange_spread: true,
    divination_card: true,
    bulk_buy: true
  }

  @all_categories [
    :cards,
    :fragments,
    :ancestor,
    :essences,
    :currency,
    :beasts,
    :map_key,
    :heist,
    :runegrafts,
    :delve,
    :sanctum,
    :maps_unique,
    :delirium_orbs,
    :oils,
    :catalysts,
    :ducats,
    :maps_special,
    :allflame_embers,
    :keepers,
    :enshrouding_crystals,
    :legacy,
    :expedition,
    :misc
  ]

  @all_categories_enabled Map.new(@all_categories, &{&1, true})

  describe "get_route_key/1" do
    test "is identical for two opportunities with the same technique and currencies" do
      a = opportunity()
      b = opportunity(%{margin_percent: 12, profit: chaos_profit(0.1), volume: 5})

      assert FlipOpportunityTablePresenter.get_route_key(a) ==
               FlipOpportunityTablePresenter.get_route_key(b)
    end

    test "differs when the technique differs" do
      a = opportunity(%{technique: :exchange_spread})
      b = opportunity(%{technique: :bulk_buy})

      refute FlipOpportunityTablePresenter.get_route_key(a) ==
               FlipOpportunityTablePresenter.get_route_key(b)
    end

    test "differs when the via currency differs" do
      a = opportunity()
      b = opportunity(%{via: [currency_amount("portal", "Portal Scroll", 10)]})

      refute FlipOpportunityTablePresenter.get_route_key(a) ==
               FlipOpportunityTablePresenter.get_route_key(b)
    end
  end

  describe "filter_by_technique/2" do
    test "keeps only opportunities whose technique is enabled" do
      spread = opportunity(%{technique: :exchange_spread})
      bulk = opportunity(%{technique: :bulk_buy})

      result =
        FlipOpportunityTablePresenter.filter_by_technique([spread, bulk], %{
          @all_enabled
          | bulk_buy: false
        })

      assert result == [spread]
    end
  end

  describe "filter_by_category/2" do
    test "keeps every opportunity when every category is enabled (default)" do
      currency =
        opportunity(%{via: [currency_amount("wisdom", "Scroll of Wisdom", 1, :currency)]})

      oil = opportunity(%{via: [currency_amount("oil", "Golden Oil", 1, :oils)]})

      result =
        FlipOpportunityTablePresenter.filter_by_category([currency, oil], @all_categories_enabled)

      assert result == [currency, oil]
    end

    test "hides opportunities whose Via category is disabled" do
      currency =
        opportunity(%{via: [currency_amount("wisdom", "Scroll of Wisdom", 1, :currency)]})

      oil = opportunity(%{via: [currency_amount("oil", "Golden Oil", 1, :oils)]})

      result =
        FlipOpportunityTablePresenter.filter_by_category(
          [currency, oil],
          %{@all_categories_enabled | oils: false}
        )

      assert result == [currency]
    end

    test "a Divination Card opportunity is categorized by the reward it gives, not the card itself" do
      # Via is [card, reward] for :divination_card (see
      # DivinationCardOpportunityFinder) -- the card leg is always :cards
      # and would be a useless filter dimension, so the category that must
      # drive filtering is the reward (the second/last Via leg).
      card_for_oil =
        opportunity(%{
          technique: :divination_card,
          via: [
            currency_amount("card", "The Doctor", 5, :cards),
            currency_amount("oil", "Golden Oil", 1, :oils)
          ]
        })

      # Disabling :cards (the card's own category) must NOT hide this row --
      # only disabling the reward's category (:oils) should.
      result_cards_disabled =
        FlipOpportunityTablePresenter.filter_by_category(
          [card_for_oil],
          %{@all_categories_enabled | cards: false}
        )

      assert result_cards_disabled == [card_for_oil]

      result_oils_disabled =
        FlipOpportunityTablePresenter.filter_by_category(
          [card_for_oil],
          %{@all_categories_enabled | oils: false}
        )

      assert result_oils_disabled == []
    end
  end

  describe "filter_by_thresholds/2" do
    test "keeps opportunities at or above every provided threshold" do
      low = opportunity(%{margin_percent: 10, volume: 5})
      high = opportunity(%{margin_percent: 90, volume: 500})

      assert FlipOpportunityTablePresenter.filter_by_thresholds([low, high], %{margin: 50}) == [
               high
             ]

      assert FlipOpportunityTablePresenter.filter_by_thresholds([low, high], %{volume: 5}) == [
               low,
               high
             ]
    end

    test "applies multiple thresholds together" do
      meets_margin_only = opportunity(%{margin_percent: 90, volume: 1})
      meets_both = opportunity(%{margin_percent: 90, volume: 100})

      result =
        FlipOpportunityTablePresenter.filter_by_thresholds([meets_margin_only, meets_both], %{
          margin: 50,
          volume: 50
        })

      assert result == [meets_both]
    end

    test "returns everything when no thresholds are provided" do
      items = [opportunity(), opportunity(%{margin_percent: 1})]

      assert FlipOpportunityTablePresenter.filter_by_thresholds(items, %{}) == items
    end
  end

  describe "filter_by_max_start_chaos/2" do
    test "passes everything through when nil" do
      items = [
        opportunity(%{start_chaos_equivalent: 1.0}),
        opportunity(%{start_chaos_equivalent: 500.0})
      ]

      assert FlipOpportunityTablePresenter.filter_by_max_start_chaos(items, nil) == items
    end

    test "keeps rows at or under the cap regardless of anchor currency, using the Chaos-normalized value" do
      cheap = opportunity(%{start_chaos_equivalent: 5.0})

      divine_anchored_but_affordable =
        opportunity(%{
          start: [currency_amount("divine", "Divine Orb", 0.05)],
          start_chaos_equivalent: 8.0
        })

      too_expensive = opportunity(%{start_chaos_equivalent: 79.0})

      result =
        FlipOpportunityTablePresenter.filter_by_max_start_chaos(
          [cheap, divine_anchored_but_affordable, too_expensive],
          10.0
        )

      assert result == [cheap, divine_anchored_but_affordable]
    end
  end

  describe "sort_opportunities/3" do
    test "sorts ascending by the requested column" do
      low = opportunity(%{profit: chaos_profit(1)})
      high = opportunity(%{profit: chaos_profit(9)})

      assert FlipOpportunityTablePresenter.sort_opportunities([high, low], :profit, :asc) == [
               low,
               high
             ]
    end

    test "sorts descending by the requested column" do
      low = opportunity(%{volume: 1})
      high = opportunity(%{volume: 9})

      assert FlipOpportunityTablePresenter.sort_opportunities([low, high], :volume, :desc) == [
               high,
               low
             ]
    end

    test "does not mutate the input list" do
      items = [opportunity(%{margin_percent: 1}), opportunity(%{margin_percent: 9})]
      original = items

      FlipOpportunityTablePresenter.sort_opportunities(items, :margin, :desc)

      assert items == original
    end
  end

  describe "partition_favorites/2" do
    test "splits opportunities into favorites and others by route key" do
      favorited = opportunity(%{technique: :exchange_spread})
      not_favorited = opportunity(%{technique: :bulk_buy})

      result =
        FlipOpportunityTablePresenter.partition_favorites(
          [favorited, not_favorited],
          MapSet.new([FlipOpportunityTablePresenter.get_route_key(favorited)])
        )

      assert result == %{favorites: [favorited], others: [not_favorited]}
    end

    test "preserves relative order within each group" do
      fav_a = opportunity(%{technique: :exchange_spread, volume: 1})

      fav_b =
        opportunity(%{
          technique: :exchange_spread,
          volume: 2,
          via: [currency_amount("x", "X", 1)]
        })

      keys =
        MapSet.new([
          FlipOpportunityTablePresenter.get_route_key(fav_a),
          FlipOpportunityTablePresenter.get_route_key(fav_b)
        ])

      result = FlipOpportunityTablePresenter.partition_favorites([fav_a, fav_b], keys)

      assert result.favorites == [fav_a, fav_b]
    end
  end

  describe "build_display_groups/2" do
    test "filters by technique and threshold, sorts, then partitions favorites -- a filtered-out favorite disappears entirely" do
      kept_favorite = opportunity(%{technique: :exchange_spread, margin_percent: 90, volume: 1})

      filtered_out_favorite =
        opportunity(%{
          technique: :bulk_buy,
          margin_percent: 5,
          volume: 1,
          via: [currency_amount("y", "Y", 1)]
        })

      non_favorite =
        opportunity(%{
          technique: :exchange_spread,
          margin_percent: 50,
          volume: 2,
          via: [currency_amount("z", "Z", 1)]
        })

      result =
        FlipOpportunityTablePresenter.build_display_groups(
          [kept_favorite, filtered_out_favorite, non_favorite],
          %{
            enabled_techniques: %{@all_enabled | bulk_buy: false},
            enabled_categories: @all_categories_enabled,
            thresholds: %{},
            max_start_chaos: nil,
            sort_column: :margin,
            sort_direction: :desc,
            favorite_route_keys:
              MapSet.new([
                FlipOpportunityTablePresenter.get_route_key(kept_favorite),
                FlipOpportunityTablePresenter.get_route_key(filtered_out_favorite)
              ])
          }
        )

      assert result.favorites == [kept_favorite]
      assert result.others == [non_favorite]
    end
  end
end
